flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -2075,10 +2154,30 @@ public void reportStats(long id, ExecutionAttemptID attemptId, CheckpointMetrics
 
     // ------------------------------------------------------------------------
 
-    private final class ScheduledTrigger implements Runnable {
+    final class ScheduledTrigger implements Runnable {
 
         @Override
         public void run() {
+            synchronized (lock) {
+                if (currentPeriodicTrigger != this) {
+                    // Another periodic trigger has been scheduled but this one
+                    // has not been force cancelled yet.
+                    return;
+                }
+
+                long checkpointInterval = getCurrentCheckpointInterval();
+                if (checkpointInterval != Long.MAX_VALUE) {
+                    nextCheckpointTriggeringRelativeTime =
+                            clock.relativeTimeMillis() + checkpointInterval;
```

Comment:
Should we use `nextCheckpointTriggeringRelativeTime += checkpointInterval` for consistency with the semantics of `scheduleAtFixedRate`()?

Then we can schedule runnable with delay = `nextCheckpointTriggeringRelativeTime - clock.relativeTimeMillis()`.